Explanations.txt

```diff
@@ -0,0 +1,92 @@
+------------------------------------------------------------------------------------------
+My project is a Minesweeper game. First, to properly remind myself of the gameplay of Minesweeper
+and to understand what the key elements of the game are I played a couple of versions of the game on my phone. 
+In my opinion the key elements of a Minesweeper are:
+1. A Game Grid - grid with cells that either contain a mine, empty, or adjacent to mines
+2. Mine placement - random distribution of mines across the grid
+3. Revealing Cells - a number, a mine, or recursively reveal surrounding cells
+4. Win/Loss conditions - all non-mines revealed/a mine revealed
+6. Input Handling - player input for the cell to be revealed. Allow flagging
+7. Timer - is present in most of the versions of the game i have tried.
+
+Also need to add Unit and Property-based tests unique to the Haskell implementaton of the game.
+
+------------------------------------------------------------------------------------------
+
+First, I setup a project. I decided to use stack, as we have been primarily using it for Homework Assignments.
+Named the project Minesweeper. From the beginning, to follow modular programming and seperation of concerns 
+principles I have decided to Each module is designed to handle a specific part of the game’s functionality, 
+ensuring high cohesion and low coupling. This approach makes the code easier to understand, test, and maintain.
+For example, the Grid module focuses solely on grid-related operations. Moved Main.hs into app folder to seperate
+the entry point of the program from the library code (src/...)
+
+------------------------------------------------------------------------------------------
+
+Second, I started working on the Game Grid. First I made a new Algebraic Data Type (ADT) 
+for cells. It has only one constructor with multiple fields. Each field corresponds to
+possible contents of the cell. A cell can either have a revealed mine, a revealed cell
+with a number of adjacent mines, a flagged cell, or an unrevealed cell.
+
+----------------------------------------------------------------------
+Next worked on random generation of grids. Explanations behind certain choices made are in the comments
+
+randomUniqueIndices :: Int -> Int -> IO [Int]
+randomUniqueIndices n maxIndex = do
+  -- replicateM to repeat a monadic action multiple times and collect results in a list
+  indices <- replicateM n (randomRIO (0, maxIndex - 1))
+  -- nub to remove duplicating indices
+  return $ take n (nub indices)
+
+wanted to do this, but then if indices are duplicated and removed due to nub there would be less than n mines
+Instead used Sets since they already handle the duplicates.
```

Comment:
I bet there's a way to repeat monadic action infinitely many time in a lazy fashion and then take enough elements from the prefix to have n unique numbers.